Content.Server/DeviceNetwork/Systems/NetworkConfiguratorSystem.cs

```diff
@@ -0,0 +1,271 @@
+
+        //UI
+        SubscribeLocalEvent<NetworkConfiguratorComponent, BoundUIClosedEvent>(OnUiClosed);
+        SubscribeLocalEvent<NetworkConfiguratorComponent, NetworkConfiguratorRemoveDeviceMessage>(OnRemoveDevice);
+        SubscribeLocalEvent<NetworkConfiguratorComponent, NetworkConfiguratorClearDevicesMessage>(OnClearDevice);
+        SubscribeLocalEvent<NetworkConfiguratorComponent, NetworkConfiguratorButtonPressedMessage>(OnConfigButtonPressed);
+    }
+
+    public void TryAddNetworkDevice(EntityUid? targetUid, EntityUid configuratorUid, EntityUid userUid,
+        NetworkConfiguratorComponent? configurator = null)
+    {
+        if (!Resolve(configuratorUid, ref configurator))
+            return;
+
+        TryAddNetworkDevice(targetUid, userUid, configurator);
+    }
+
+    private void TryAddNetworkDevice(EntityUid? targetUid, EntityUid userUid, NetworkConfiguratorComponent configurator, DeviceNetworkComponent? device = null)
+    {
+        if (!targetUid.HasValue || !Resolve(targetUid.Value, ref device))
+            return;
+
+        if (string.IsNullOrEmpty(device.Address))
+        {
+            _popupSystem.PopupCursor(Loc.GetString("network-configurator-device-failed", ("device", targetUid)), Filter.Entities(userUid));
+            return;
+        }
+
+        if (configurator.Devices.ContainsValue(targetUid.Value))
+        {
+            _popupSystem.PopupCursor(Loc.GetString("network-configurator-device-already-saved", ("device", targetUid)), Filter.Entities(userUid));
+            return;
+        }
+
+        configurator.Devices.Add(device.Address, targetUid.Value);
+        _popupSystem.PopupCursor(Loc.GetString("network-configurator-device-saved", ("address", device.Address), ("device", targetUid)),
+            Filter.Entities(userUid));
+
+        UpdateUiState(configurator.Owner, configurator);
+    }
+
+    private bool AccessCheck(EntityUid target, EntityUid? user, NetworkConfiguratorComponent component)
+    {
+        if (!TryComp(target, out AccessReaderComponent? reader) || user == null)
+            return false;
+
+        if (_accessSystem.IsAllowed(reader, user.Value))
+            return true;
+
+        SoundSystem.Play(Filter.Pvs(user.Value), component.SoundNoAccess.GetSound(), target, AudioParams.Default.WithVolume(-2f).WithPitchScale(1.2f));
+        _popupSystem.PopupEntity(Loc.GetString("network-configurator-device-access-denied"), target, Filter.Entities(user.Value));
+
+        return false;
+    }
+
+
+    #region Interactions
+
+    /// <summary>
+    /// Shows the saved devices ui. It can be used to view and remove saved devices
+    /// </summary>
+    private void OnUsedInHand(EntityUid uid, NetworkConfiguratorComponent component, UseInHandEvent args)
+    {
+        if (!TryComp(args.User, out ActorComponent? actor))
+            return;
+
+        _uiSystem.GetUiOrNull(uid, NetworkConfiguratorUiKey.List)?.Open(actor.PlayerSession);
+        UpdateUiState(uid, component);
+    }
```

Comment:
this should use `ActivatableUI` in YAML rather than reimplementing it. that also has more robust checks for whether it can be used